<a href="https://colab.research.google.com/github/AshKatzEm/Movie_recommender/blob/main/Cleaning_and_Exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Before we can model our data, we need see what exactly we have, make sure it's clean, and remove anything we don't want.


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
movies = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/movies.csv') 

genomeTags = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/genome-tags.csv')  

genomeScores = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/genome-scores.csv') 

ratings = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/ratings.csv') 

tags = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/tags.csv') 

links = pd.read_csv('/content/drive/MyDrive/Movie_recommender/Data/links.csv') 

# pd.set_option('display.max_rows', 10000)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.width', None)
# pd.set_option('display.max_colwidth', None)


In [3]:
movies

movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres  
0      Adventure|Animation|Children|Comedy|Fantasy  
1                       Adventure|Children|Fantasy  
2                                   Comedy|Romance  
3                             Comedy|Drama|Romance  
4                                           Comedy  
...                                            ...  
58093                           (no genres listed)  
58094                                       Comedy  
58095                                        Drama  
58096                Adventure|Drama|Horror|Sci-Fi  
58097                           Action|Crime|Drama  

[58098 rows x 3 columns]

In [4]:
genomeTags

tagId           tag
0         1           007
1         2  007 (series)
2         3  18th century
3         4         1920s
4         5         1930s
...     ...           ...
1123   1124       writing
1124   1125         wuxia
1125   1126          wwii
1126   1127        zombie
1127   1128       zombies

[1128 rows x 2 columns]

In [5]:
genomeScores

movieId  tagId  relevance
0               1      1    0.02900
1               1      2    0.02375
2               1      3    0.05425
3               1      4    0.06875
4               1      5    0.16000
...           ...    ...        ...
14862523   187595   1124    0.10700
14862524   187595   1125    0.05850
14862525   187595   1126    0.03800
14862526   187595   1127    0.10225
14862527   187595   1128    0.02925

[14862528 rows x 3 columns]

In [6]:
ratings

userId  movieId  rating   timestamp
0              1      307     3.5  1256677221
1              1      481     3.5  1256677456
2              1     1091     1.5  1256677471
3              1     1257     4.5  1256677460
4              1     1449     4.5  1256677264
...          ...      ...     ...         ...
27753439  283228     8542     4.5  1379882795
27753440  283228     8712     4.5  1379882751
27753441  283228    34405     4.5  1379882889
27753442  283228    44761     4.5  1354159524
27753443  283228    54286     4.5  1354159718

[27753444 rows x 4 columns]

In [7]:
tags

userId  movieId                 tag   timestamp
0            14      110                epic  1443148538
1            14      110            Medieval  1443148532
2            14      260              sci-fi  1442169410
3            14      260        space action  1442169421
4            14      318        imdb top 250  1442615195
...         ...      ...                 ...         ...
1108992  283206    73017                 fun  1264379059
1108993  283206    73017  homoerotic subtext  1264379058
1108994  283206    73017              pacing  1264379058
1108995  283206    73017                plot  1264379058
1108996  283221    49651  Sylvester Stallone  1168346830

[1108997 rows x 4 columns]

In [8]:
links

movieId   imdbId    tmdbId
0            1   114709     862.0
1            2   113497    8844.0
2            3   113228   15602.0
3            4   114885   31357.0
4            5   113041   11862.0
...        ...      ...       ...
58093   193876    38566   78251.0
58094   193878  1754787   87558.0
58095   193880  5847740  422666.0
58096   193882  4453756  454439.0
58097   193886  7606620  540871.0

[58098 rows x 3 columns]

## Currently, the "genres" column of the "movies" dataframe combines all of a movie's genres as a single string. Let's see if we can split it up

In [9]:
movies = movies.join(movies.genres.str.split("|", expand = True))
movies

movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres                   0  \
0      Adventure|Animation|Children|Comedy|Fantasy           Adventure   
1                       Adventure|Children|Fantasy           Adventure   
2                                   Comedy|Romance              Comedy   
3                             Comedy|Drama|Romance              Comedy   
4                                           Comedy              Comedy   
...                                            ...                 ...   
58093                           (no genres listed)  (no genres listed)   
58094                                       Comedy              Comedy   
58095                                        Drama               Drama   
58096                Adventure|Drama|Horror|Sci-Fi           Adventure   
58097                           Action|Crime|Drama              Action   

               1         2       3        4     5     6     7     8     9  
0      Animation  Children  Comedy  Fantasy  None  None  None  None  None  
1       Children   Fantasy    None     None  None  None  None  None  None  
2        Romance      None    None     None  None  None  None  None  None  
3          Drama   Romance    None     None  None  None  None  None  None  
4           None      None    None     None  None  None  None  None  None  
...          ...       ...     ...      ...   ...   ...   ...   ...   ...  
58093       None      None    None     None  None  None  None  None  None  
58094       None      None    None     None  None  None  None  None  None  
58095       None      None    None     None  None  None  None  None  None  
58096      Drama    Horror  Sci-Fi     None  None  None  None  None  None  
58097      Crime     Drama    None     None  None  None  None  None  None  

[58098 rows x 13 columns]

ok, those column names are not working

In [10]:
movies = movies.rename(columns = {0:'genre1', 1:'genre2',2:'genre3',3:'genre4',4:'genre5',5:'genre6', 6: 'genre7',7:'genre8',8:'genre9',9:'genre10'})
movies

movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres              genre1  \
0      Adventure|Animation|Children|Comedy|Fantasy           Adventure   
1                       Adventure|Children|Fantasy           Adventure   
2                                   Comedy|Romance              Comedy   
3                             Comedy|Drama|Romance              Comedy   
4                                           Comedy              Comedy   
...                                            ...                 ...   
58093                           (no genres listed)  (no genres listed)   
58094                                       Comedy              Comedy   
58095                                        Drama               Drama   
58096                Adventure|Drama|Horror|Sci-Fi           Adventure   
58097                           Action|Crime|Drama              Action   

          genre2    genre3  genre4   genre5 genre6 genre7 genre8 genre9  \
0      Animation  Children  Comedy  Fantasy   None   None   None   None   
1       Children   Fantasy    None     None   None   None   None   None   
2        Romance      None    None     None   None   None   None   None   
3          Drama   Romance    None     None   None   None   None   None   
4           None      None    None     None   None   None   None   None   
...          ...       ...     ...      ...    ...    ...    ...    ...   
58093       None      None    None     None   None   None   None   None   
58094       None      None    None     None   None   None   None   None   
58095       None      None    None     None   None   None   None   None   
58096      Drama    Horror  Sci-Fi     None   None   None   None   None   
58097      Crime     Drama    None     None   None   None   None   None   

      genre10  
0        None  
1        None  
2        None  
3        None  
4        None  
...       ...  
58093    None  
58094    None  
58095    None  
58096    None  
58097    None  

[58098 rows x 13 columns]

## This method is actually pretty messy. Let's try one-hot encoding each genre. First, let's see how many genres we have

In [11]:
movies['genre1'].unique()

array(['Adventure', 'Comedy', 'Action', 'Drama', 'Crime', 'Children',
       'Mystery', 'Documentary', 'Animation', 'Thriller', 'Horror',
       'Fantasy', 'Western', 'Film-Noir', 'Romance', 'War', 'Sci-Fi',
       'Musical', 'IMAX', '(no genres listed)'], dtype=object)

In [12]:
movies['genre2'].unique()

array(['Animation', 'Children', 'Romance', 'Drama', None, 'Crime',
       'Adventure', 'Horror', 'Comedy', 'Sci-Fi', 'IMAX', 'War',
       'Thriller', 'Mystery', 'Film-Noir', 'Fantasy', 'Musical',
       'Documentary', 'Western'], dtype=object)

## It looks like every possible genre appears in the first genre column. Let's use that value list to create a column for each genre. 

In [13]:
for i in movies['genre1'].unique():
  movies[i] = 0

movies

movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres              genre1  \
0      Adventure|Animation|Children|Comedy|Fantasy           Adventure   
1                       Adventure|Children|Fantasy           Adventure   
2                                   Comedy|Romance              Comedy   
3                             Comedy|Drama|Romance              Comedy   
4                                           Comedy              Comedy   
...                                            ...                 ...   
58093                           (no genres listed)  (no genres listed)   
58094                                       Comedy              Comedy   
58095                                        Drama               Drama   
58096                Adventure|Drama|Horror|Sci-Fi           Adventure   
58097                           Action|Crime|Drama              Action   

          genre2    genre3  genre4   genre5 genre6 genre7  ... Horror Fantasy  \
0      Animation  Children  Comedy  Fantasy   None   None  ...      0       0   
1       Children   Fantasy    None     None   None   None  ...      0       0   
2        Romance      None    None     None   None   None  ...      0       0   
3          Drama   Romance    None     None   None   None  ...      0       0   
4           None      None    None     None   None   None  ...      0       0   
...          ...       ...     ...      ...    ...    ...  ...    ...     ...   
58093       None      None    None     None   None   None  ...      0       0   
58094       None      None    None     None   None   None  ...      0       0   
58095       None      None    None     None   None   None  ...      0       0   
58096      Drama    Horror  Sci-Fi     None   None   None  ...      0       0   
58097      Crime     Drama    None     None   None   None  ...      0       0   

      Western  Film-Noir  Romance  War  Sci-Fi  Musical  IMAX  \
0           0          0        0    0       0        0     0   
1           0          0        0    0       0        0     0   
2           0          0        0    0       0        0     0   
3           0          0        0    0       0        0     0   
4           0          0        0    0       0        0     0   
...       ...        ...      ...  ...     ...      ...   ...   
58093       0          0        0    0       0        0     0   
58094       0          0        0    0       0        0     0   
58095       0          0        0    0       0        0     0   
58096       0          0        0    0       0        0     0   
58097       0          0        0    0       0        0     0   

       (no genres listed)  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
58093                   0  
58094                   0  
58095                   0  
58096                   0  
58097                   0  

[58098 rows x 33 columns]

In [14]:
genres = movies['genre1'].unique()
for i in genres:
  movies[i] = 0

movies



movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres              genre1  \
0      Adventure|Animation|Children|Comedy|Fantasy           Adventure   
1                       Adventure|Children|Fantasy           Adventure   
2                                   Comedy|Romance              Comedy   
3                             Comedy|Drama|Romance              Comedy   
4                                           Comedy              Comedy   
...                                            ...                 ...   
58093                           (no genres listed)  (no genres listed)   
58094                                       Comedy              Comedy   
58095                                        Drama               Drama   
58096                Adventure|Drama|Horror|Sci-Fi           Adventure   
58097                           Action|Crime|Drama              Action   

          genre2    genre3  genre4   genre5 genre6 genre7  ... Horror Fantasy  \
0      Animation  Children  Comedy  Fantasy   None   None  ...      0       0   
1       Children   Fantasy    None     None   None   None  ...      0       0   
2        Romance      None    None     None   None   None  ...      0       0   
3          Drama   Romance    None     None   None   None  ...      0       0   
4           None      None    None     None   None   None  ...      0       0   
...          ...       ...     ...      ...    ...    ...  ...    ...     ...   
58093       None      None    None     None   None   None  ...      0       0   
58094       None      None    None     None   None   None  ...      0       0   
58095       None      None    None     None   None   None  ...      0       0   
58096      Drama    Horror  Sci-Fi     None   None   None  ...      0       0   
58097      Crime     Drama    None     None   None   None  ...      0       0   

      Western  Film-Noir  Romance  War  Sci-Fi  Musical  IMAX  \
0           0          0        0    0       0        0     0   
1           0          0        0    0       0        0     0   
2           0          0        0    0       0        0     0   
3           0          0        0    0       0        0     0   
4           0          0        0    0       0        0     0   
...       ...        ...      ...  ...     ...      ...   ...   
58093       0          0        0    0       0        0     0   
58094       0          0        0    0       0        0     0   
58095       0          0        0    0       0        0     0   
58096       0          0        0    0       0        0     0   
58097       0          0        0    0       0        0     0   

       (no genres listed)  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
58093                   0  
58094                   0  
58095                   0  
58096                   0  
58097                   0  

[58098 rows x 33 columns]

In [15]:
for i in movies.index: # rows of the df
  for j in genres: # the array of possible values, not the df column
    if j in movies.genres[i]: # if a genre is contained in a movies genres
      movies[j][i] = 1 # store a 1 in that genres individual column
movies

<ipython-input-15-cbc86a36363d>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies[j][i] = 1 # store a 1 in that genres individual column


movieId                                    title  \
0            1                         Toy Story (1995)   
1            2                           Jumanji (1995)   
2            3                  Grumpier Old Men (1995)   
3            4                 Waiting to Exhale (1995)   
4            5       Father of the Bride Part II (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                                            genres              genre1  \
0      Adventure|Animation|Children|Comedy|Fantasy           Adventure   
1                       Adventure|Children|Fantasy           Adventure   
2                                   Comedy|Romance              Comedy   
3                             Comedy|Drama|Romance              Comedy   
4                                           Comedy              Comedy   
...                                            ...                 ...   
58093                           (no genres listed)  (no genres listed)   
58094                                       Comedy              Comedy   
58095                                        Drama               Drama   
58096                Adventure|Drama|Horror|Sci-Fi           Adventure   
58097                           Action|Crime|Drama              Action   

          genre2    genre3  genre4   genre5 genre6 genre7  ... Horror Fantasy  \
0      Animation  Children  Comedy  Fantasy   None   None  ...      0       1   
1       Children   Fantasy    None     None   None   None  ...      0       1   
2        Romance      None    None     None   None   None  ...      0       0   
3          Drama   Romance    None     None   None   None  ...      0       0   
4           None      None    None     None   None   None  ...      0       0   
...          ...       ...     ...      ...    ...    ...  ...    ...     ...   
58093       None      None    None     None   None   None  ...      0       0   
58094       None      None    None     None   None   None  ...      0       0   
58095       None      None    None     None   None   None  ...      0       0   
58096      Drama    Horror  Sci-Fi     None   None   None  ...      1       0   
58097      Crime     Drama    None     None   None   None  ...      0       0   

      Western  Film-Noir  Romance  War  Sci-Fi  Musical  IMAX  \
0           0          0        0    0       0        0     0   
1           0          0        0    0       0        0     0   
2           0          0        1    0       0        0     0   
3           0          0        1    0       0        0     0   
4           0          0        0    0       0        0     0   
...       ...        ...      ...  ...     ...      ...   ...   
58093       0          0        0    0       0        0     0   
58094       0          0        0    0       0        0     0   
58095       0          0        0    0       0        0     0   
58096       0          0        0    0       1        0     0   
58097       0          0        0    0       0        0     0   

       (no genres listed)  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
58093                   1  
58094                   0  
58095                   0  
58096                   0  
58097                   0  

[58098 rows x 33 columns]

## That looks like it worked, so let's clean up the previous attempt


In [16]:
movies = movies.drop(labels=['genres', 'genre1', 'genre2', 'genre3','genre4','genre5','genre6', 'genre7','genre8','genre9','genre10'], axis=1)

In [17]:
movies

movieId                                    title  Adventure  Comedy  \
0            1                         Toy Story (1995)          1       1   
1            2                           Jumanji (1995)          1       0   
2            3                  Grumpier Old Men (1995)          0       1   
3            4                 Waiting to Exhale (1995)          0       1   
4            5       Father of the Bride Part II (1995)          0       1   
...        ...                                      ...        ...     ...   
58093   193876                  The Great Glinka (1946)          0       0   
58094   193878  Les tribulations d'une caissière (2011)          0       1   
58095   193880                 Her Name Was Mumu (2016)          0       0   
58096   193882                             Flora (2017)          1       0   
58097   193886                              Leal (2018)          0       0   

       Action  Drama  Crime  Children  Mystery  Documentary  ...  Horror  \
0           0      0      0         1        0            0  ...       0   
1           0      0      0         1        0            0  ...       0   
2           0      0      0         0        0            0  ...       0   
3           0      1      0         0        0            0  ...       0   
4           0      0      0         0        0            0  ...       0   
...       ...    ...    ...       ...      ...          ...  ...     ...   
58093       0      0      0         0        0            0  ...       0   
58094       0      0      0         0        0            0  ...       0   
58095       0      1      0         0        0            0  ...       0   
58096       0      1      0         0        0            0  ...       1   
58097       1      1      1         0        0            0  ...       0   

       Fantasy  Western  Film-Noir  Romance  War  Sci-Fi  Musical  IMAX  \
0            1        0          0        0    0       0        0     0   
1            1        0          0        0    0       0        0     0   
2            0        0          0        1    0       0        0     0   
3            0        0          0        1    0       0        0     0   
4            0        0          0        0    0       0        0     0   
...        ...      ...        ...      ...  ...     ...      ...   ...   
58093        0        0          0        0    0       0        0     0   
58094        0        0          0        0    0       0        0     0   
58095        0        0          0        0    0       0        0     0   
58096        0        0          0        0    0       1        0     0   
58097        0        0          0        0    0       0        0     0   

       (no genres listed)  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
58093                   1  
58094                   0  
58095                   0  
58096                   0  
58097                   0  

[58098 rows x 22 columns]

# Ok, so let's move on. 
## In our final dataset, I think we only need the "Id"s. When a user enters a a tag, or a movie name, we can do a quick conversion with the genome or movies dataframe. In fact, let's create a "conversion" dataframe.

In [18]:
conversion = pd.DataFrame(data = movies.title)
conversion['movieId'] = movies.movieId
conversion['tag'] = genomeTags.tag
conversion['tagId'] = genomeTags.tagId
conversion


title  movieId           tag  tagId
0                             Toy Story (1995)        1           007    1.0
1                               Jumanji (1995)        2  007 (series)    2.0
2                      Grumpier Old Men (1995)        3  18th century    3.0
3                     Waiting to Exhale (1995)        4         1920s    4.0
4           Father of the Bride Part II (1995)        5         1930s    5.0
...                                        ...      ...           ...    ...
58093                  The Great Glinka (1946)   193876           NaN    NaN
58094  Les tribulations d'une caissière (2011)   193878           NaN    NaN
58095                 Her Name Was Mumu (2016)   193880           NaN    NaN
58096                             Flora (2017)   193882           NaN    NaN
58097                              Leal (2018)   193886           NaN    NaN

[58098 rows x 4 columns]

## That looks like it will work, albeit a little akward. Now, let's start making our final dataset and then we'll remove the non-id columns

In [19]:
RatingsMerge = pd.merge(
    ratings,
    tags,
    how="outer",
    on=["userId", 'movieId'],
    left_on=None,
    right_on=None,
    left_index=False,
    right_index=False,
    sort=False,
    suffixes=("_x", "_y"),
    copy=False,
    indicator=False,
    validate=None,
)
RatingsMerge

userId  movieId  rating   timestamp_x        tag   timestamp_y
0              1      307     3.5  1.256677e+09        NaN           NaN
1              1      481     3.5  1.256677e+09        NaN           NaN
2              1     1091     1.5  1.256677e+09        NaN           NaN
3              1     1257     4.5  1.256677e+09        NaN           NaN
4              1     1449     4.5  1.256677e+09        NaN           NaN
...          ...      ...     ...           ...        ...           ...
28641996  283184     6808     NaN           NaN      nazis  1.368473e+09
28641997  283184     8781     NaN           NaN   politics  1.368473e+09
28641998  283184    53121     NaN           NaN  franchise  1.368473e+09
28641999  283184    55290     NaN           NaN     police  1.368473e+09
28642000  283184    64622     NaN           NaN      nazis  1.368473e+09

[28642001 rows x 6 columns]

In [20]:
print(len(tags))
print(len(RatingsMerge))
print(RatingsMerge.tag.isna().sum(), "+", len(tags), '=', RatingsMerge.tag.isna().sum() + len(tags) )

1108997
28642001
27533020 + 1108997 = 28642017


## so that merge seems to have worked, numerically at least

In [30]:
print(len(ratings))
print(len(RatingsMerge))
print(RatingsMerge.rating.isna().sum(), "+", len(ratings), '=', RatingsMerge.rating.isna().sum() + len(ratings) )
print(ratings.rating.isna().sum())
print(RatingsMerge.userId.unique())

27753444
28642001
278505 + 27753444 = 28031949
0
[     1      2      3 ... 283226 283227 283228]


In [24]:
tagmerge = pd.merge(
    tags,
    ratings,
    how="left",
    on=["userId", 'movieId'],
    left_on=None,
    right_on=None,
    left_index=False,
    right_index=False,
    sort=False,
    suffixes=("_x", "_y"),
    copy=False,
    indicator=False,
    validate=None,
)
tagmerge

userId  movieId                 tag  timestamp_x  rating  \
0            14      110                epic   1443148538     4.5   
1            14      110            Medieval   1443148532     4.5   
2            14      260              sci-fi   1442169410     NaN   
3            14      260        space action   1442169421     NaN   
4            14      318        imdb top 250   1442615195     4.5   
...         ...      ...                 ...          ...     ...   
1108992  283206    73017                 fun   1264379059     4.0   
1108993  283206    73017  homoerotic subtext   1264379058     4.0   
1108994  283206    73017              pacing   1264379058     4.0   
1108995  283206    73017                plot   1264379058     4.0   
1108996  283221    49651  Sylvester Stallone   1168346830     4.0   

          timestamp_y  
0        1.442170e+09  
1        1.442170e+09  
2                 NaN  
3                 NaN  
4        1.442169e+09  
...               ...  
1108992  1.264379e+09  
1108993  1.264379e+09  
1108994  1.264379e+09  
1108995  1.264379e+09  
1108996  1.168347e+09  

[1108997 rows x 6 columns]

In [33]:
print(tagmerge.rating.isna().sum())
print(len(tagmerge.userId.unique()))

278505
19325


## That seems to be ok, let's try to roll with this

In [26]:
tagmerge = pd.merge(
    tagmerge,
    movies,
    how="inner",
    on=['movieId'],
    left_on=None,
    right_on=None,
    left_index=False,
    right_index=False,
    sort=False,
    suffixes=("_x", "_y"),
    copy=False,
    indicator=False,
    validate=None,
)
tagmerge

userId  movieId                   tag  timestamp_x  rating  \
0            14      110                  epic   1443148538     4.5   
1            14      110              Medieval   1443148532     4.5   
2           815      110             overrated   1150006110     2.5   
3          2577      110  Oscar (Best Picture)   1378324225     4.5   
4          3086      110                  epic   1463675332     5.0   
...         ...      ...                   ...          ...     ...   
1108992  282119   164453              engineer   1475084794     4.0   
1108993  282119   164453       soviet classics   1475084778     4.0   
1108994  283000    91960         Amazingly Bad   1337484461     0.5   
1108995  283000    91960       Effeminate Tree   1337484484     0.5   
1108996  283000    91960      Riff Trax Fodder   1337484523     0.5   

          timestamp_y                             title  Adventure  Comedy  \
0        1.442170e+09                 Braveheart (1995)          0       0   
1        1.442170e+09                 Braveheart (1995)          0       0   
2        1.053219e+09                 Braveheart (1995)          0       0   
3        1.378324e+09                 Braveheart (1995)          0       0   
4        1.462977e+09                 Braveheart (1995)          0       0   
...               ...                               ...        ...     ...   
1108992  1.475085e+09  Failure of Engineer Garin (1973)          1       0   
1108993  1.475085e+09  Failure of Engineer Garin (1973)          1       0   
1108994  1.335057e+09  Magic Christmas Tree, The (1964)          0       1   
1108995  1.335057e+09  Magic Christmas Tree, The (1964)          0       1   
1108996  1.335057e+09  Magic Christmas Tree, The (1964)          0       1   

         Action  ...  Horror  Fantasy  Western  Film-Noir  Romance  War  \
0             1  ...       0        0        0          0        0    1   
1             1  ...       0        0        0          0        0    1   
2             1  ...       0        0        0          0        0    1   
3             1  ...       0        0        0          0        0    1   
4             1  ...       0        0        0          0        0    1   
...         ...  ...     ...      ...      ...        ...      ...  ...   
1108992       0  ...       0        0        0          0        0    0   
1108993       0  ...       0        0        0          0        0    0   
1108994       0  ...       0        1        0          0        0    0   
1108995       0  ...       0        1        0          0        0    0   
1108996       0  ...       0        1        0          0        0    0   

         Sci-Fi  Musical  IMAX  (no genres listed)  
0             0        0     0                   0  
1             0        0     0                   0  
2             0        0     0                   0  
3             0        0     0                   0  
4             0        0     0                   0  
...         ...      ...   ...                 ...  
1108992       1        0     0                   0  
1108993       1        0     0                   0  
1108994       0        0     0                   0  
1108995       0        0     0                   0  
1108996       0        0     0                   0  

[1108997 rows x 27 columns]

# Let's just leave the "title" and "tag" features in, we can always drop it later if neccesary.
## We will conclude with two datasets
* ## The original ratings dataset, which can be useful in that it has 27,753,444 ratings from 283,228 users
* ## A rich dataset with far more features attached to 850,000 ratings from 19,325 users

In [34]:
tagmerge.to_csv('/content/drive/MyDrive/Movie_recommender/Data/RichMovieDataset.csv', index = False)
conversion.to_csv('/content/drive/MyDrive/Movie_recommender/Data/conversionTable.csv', index = False)